[mod] theme/simple: refactor codebase

inetol opened the pull request:
This is one of various PR to refactor the simple theme internally.

TypeScript is a superset of JavaScript, converting the entire theme to TypeScript allows us to receive much more feedback on possible issues made in package updates or our own typos, furthermore, it allows to transpile properly to lower specs. This PR couldn't be done in smaller commits, a lot of work needed to make everything *work properly*:

From https://github.com/searxng/searxng/pull/4988:
- A browser baseline has been set that requires minimum **Chromium 93, Firefox 92 and Safari 15** (proper visuals/operation on older browser versions is not guaranteed)
- LightningCSS now handles minification and prefix creation for CSS.
- All hardcoded polyfills and support for previous browser baseline versions have been removed.
- Convert codebase to TypeScript.
- Convert IIFE to ESM, handling globals with IIFE is cumbersome, ESM is the standard for virtually any use of JS nowadays.
- Vite now builds the theme without the need for `vite-plugin-static-copy`.
- `searxng.ready` now accepts an array of conditions for the callback to be executed.
- Replace `leaflet` with `ol` as there were some issues with proper Vite bundling.
- Merged `head` with `main` script, as head was too small now.
- Add `assertElement` to properly check the existence of critical DOM elements.
- `searxng.on` renamed to `searxng.listen` with some handling improvements.

From https://github.com/searxng/searxng/pull/5073:
- Lay the foundation for loading scripts granularly depending on the endpoint it's on.
- Remove vendor specific prefixes as there are now managed by browserslist and LightningCSS.
- Enabled quite a few rules in Biome that don't come in recommended to better catch issues and improve consistency.

---

Visual from baseline browsers:

<details>
<summary>Chromium 93</summary>

Search:
![chrome93-search](https://github.com/user-attachments/assets/95e35e95-3b64-4111-9e3f-92c4bad155fd)

Images: (2nd review; image fallback was not working properly)
![chrome93-images](https://github.com/user-attachments/assets/7c44395d-7f4e-4760-9e1f-3c165ebb1deb)

Maps:
![chrome93-maps](https://github.com/user-attachments/assets/d34cba7e-69d4-46bb-8c6b-9391f019cd9c)

Preferences:
![chrome93-preferences](https://github.com/user-attachments/assets/00bbe77a-c7b0-4ffb-8ee9-a49f30d7bfe4)

</details>

<details>
<summary>Firefox 92</summary>

Search:
![firefox92-search](https://github.com/user-attachments/assets/fff25578-4689-43cc-b7d3-709688954451)

Images:
![firefox92-images](https://github.com/user-attachments/assets/4e216f98-9c09-4283-b023-fd9d777e10eb)

Maps:
![firefox92-maps](https://github.com/user-attachments/assets/ba4b43ab-d680-445c-8af9-761793d5d4ab)

Preferences:
![firefox92-preferences](https://github.com/user-attachments/assets/710d0db7-7730-4ab0-996e-7752ba5c6aa8)

</details>

<details>
<summary>Safari 15</summary>

Search:
![safari15-search](https://github.com/user-attachments/assets/6f6b6a2d-d424-4afc-a26f-a43fc1b8edad)

Images:
![safari15-images](https://github.com/user-attachments/assets/4ce258ab-a20f-4d9b-bc4f-316db643bd38)

Maps:
![safari15-maps](https://github.com/user-attachments/assets/0a658b15-9510-4a8d-bf9d-a61fe42f8975)

Preferences:
![safari15-preferences](https://github.com/user-attachments/assets/3c1f5e4c-8ff3-443e-a2d4-d280b3079603)

</details>

inetol committed:
[mod] theme/simple: migrate codebase to TypeScript

TypeScript is a superset of JavaScript, converting the entire theme to
TypeScript allows us to receive much more feedback on possible issues made in
package updates or our own typos, furthermore, it allows to transpile properly
to lower specs. This PR couldn't be done in smaller commits, a lot of work
needed to make everything *work properly*:

- A browser baseline has been set that requires minimum **Chromium 93, Firefox
  92 and Safari 15** (proper visuals/operation on older browser versions is not
  guaranteed)
- LightningCSS now handles minification and prefix creation for CSS.
- All hardcoded polyfills and support for previous browser baseline versions
  have been removed.
- Convert codebase to TypeScript.
- Convert IIFE to ESM, handling globals with IIFE is cumbersome, ESM is the
  standard for virtually any use of JS nowadays.
- Vite now builds the theme without the need for `vite-plugin-static-copy`.
- `searxng.ready` now accepts an array of conditions for the callback to be
  executed.
- Replace `leaflet` with `ol` as there were some issues with proper Vite
  bundling.
- Merged `head` with `main` script, as head was too small now.
- Add `assertElement` to properly check the existence of critical DOM elements.
- `searxng.on` renamed to `searxng.listen` with some handling improvements.

inetol committed:
[enh] theme/simple: custom router

Lay the foundation for loading scripts granularly depending on the endpoint it's
on.

Remove vendor specific prefixes as there are now managed by browserslist and
LightningCSS.

Enabled quite a few rules in Biome that don't come in recommended to better
catch issues and improve consistency.

Related:

- https://github.com/searxng/searxng/pull/5073#discussion_r2256037965
- https://github.com/searxng/searxng/pull/5073#discussion_r2256057100

inetol committed:
[enh] theme/simple: package.json revision

Set minor versioning for most of the packages that iterate fast or we know won't
cause problems, and fixed versioning for the rest.

Packages going into bundles should be placed in "dependencies".

The inspection of prod bundles is necessary. Although it does not make a lot of
sense right now, it will be useful in later PR and will give us a reference to
start with.

inetol committed:
[fix] theme/simple: broken highlightResult

From `mod-simple-strict`

return42 commented on Makefile:
```diff
@@ -84,7 +84,7 @@ MANAGE += py.build py.clean
 MANAGE += format.python
 MANAGE += test.yamllint test.pylint test.black test.pybabel test.unit test.coverage test.robot test.rst test.clean test.themes test.types.dev test.types.ci
-MANAGE += themes.all themes.simple themes.fix themes.lint themes.test
+MANAGE += themes.all themes.simple themes.simple.analyze themes.simple.analyse themes.fix themes.lint themes.test
```
`themes.simple.analyze` is listed twice / I fix it in my rebase ..

inetol replied:
Both are aliased on US and UK english because I tend to write both, but it's a dirty approach so you can remove this

return42 replied:
Oops haven't seen its an alias .. I removed `themes.simple.analyse` [diff](https://github.com/searxng/searxng/compare/e09f3561724faf991d947a45f9c733896a40da07..e0a4ee96f10e26ac8b23a97fe878444dffcebda2) .. since Makefile and `themes.simple.analyze` are not more than wrappers around `VITE_BUNDLE_ANALYZE=true npm run build` we should not add more names.

return42 commented on client/simple/src/js/main/results.ts:
```diff
@@ -0,0 +1,178 @@
+  imgLoader.src = imgSrc;
+};
+
+const imageThumbnails: NodeListOf<HTMLImageElement> =
+  document.querySelectorAll<HTMLImageElement>("#urls img.image_thumbnail");
+for (const thumbnail of imageThumbnails) {
```
Not familiar with common formatting rules in the JS & TS world: Shouldn't there be a blank line between the constant and the loop?

I took this as an opportunity to test the workflow for client development .. from the docs: 

----
<img width="887" height="241" alt="grafik" src="https://github.com/user-attachments/assets/30b695ca-51b2-4df0-92bb-5c1640a0b68e" />

----


To this end, I incorporated formatting errors.

```diff
diff --git a/client/simple/src/js/main/results.ts b/client/simple/src/js/main/results.ts
index a0fa89ab3..97b0046c4 100644
--- a/client/simple/src/js/main/results.ts
+++ b/client/simple/src/js/main/results.ts
@@ -13,9 +13,13 @@ const loadImage = (imgSrc: string, onSuccess: () => void): void => {
   imgLoader.src = imgSrc;
 };
 
-const imageThumbnails: NodeListOf<HTMLImageElement> =
+  const imageThumbnails: NodeListOf<HTMLImageElement> =
   document.querySelectorAll<HTMLImageElement>("#urls img.image_thumbnail");
-for (const thumbnail of imageThumbnails) {
+
+
+
+
+  for (const thumbnail of imageThumbnails) {
   if (thumbnail.complete && thumbnail.naturalWidth === 0) {
     thumbnail.src = `${settings.theme_static_path}/img/img_load_error.svg`;
   }
```

When I now run `make themes.simple` no issues are reported .. but when I run `make themes.fix` the diff from above is formatted into what I would expected:

```diff
diff --git a/client/simple/src/js/main/results.ts b/client/simple/src/js/main/results.ts
index a0fa89ab3..41c160849 100644
--- a/client/simple/src/js/main/results.ts
+++ b/client/simple/src/js/main/results.ts
@@ -15,6 +15,7 @@ const loadImage = (imgSrc: string, onSuccess: () => void): void => {
 
 const imageThumbnails: NodeListOf<HTMLImageElement> =
   document.querySelectorAll<HTMLImageElement>("#urls img.image_thumbnail");
+
 for (const thumbnail of imageThumbnails) {
   if (thumbnail.complete && thumbnail.naturalWidth === 0) {
     thumbnail.src = `${settings.theme_static_path}/img/img_load_error.svg`;
```

inetol replied:
> Not familiar with common formatting rules in the JS & TS world: Shouldn't there be a blank line between the constant and the loop?

Depends on what writing style you have, since the constant is used immediately, I put it together. We can enforce some basic styling rules on https://biomejs.dev/linter/javascript/rules/#style btw

return42 replied:
>  We can enforce some basic styling rules on https://biomejs.dev/linter/javascript/rules/#style btw

Since we rewrote all sources in this PR, I think it is an good opportunity to define some basic styling rules .. nothing special, just what is commonly known as "best practice".

inetol replied:
I've added more style rules in ef34245581156c47ccf1cc029474807b5cbb6523

All (almost) best practices and our opinionated style are already enforced with Biome.

return42 commented on client/simple/package.json:
```diff
@@ -1,36 +1,51 @@
   "license": "AGPL-3.0",
   "type": "module",
   "scripts": {
     "build": "npm run build:icons && npm run build:vite",
```
with my [comment](https://github.com/searxng/searxng/pull/5073/files#r2255949344) in mind: isn't it better we *lint* before build?

We could implement a *lint* here ... or in the `lib_sxng_vite.sh` .. or in the `lib_sxng_thems.sh` .. whatever you think is the best place.

Since only a few developers have worked on the client so far, we are free to design the developer workflows as we see fit and can make changes without looking back.

We should design them in the same way as is customary in JS/TS development.

The shell scripts (and the Makefile) are just wrappers that should provide simplified access to the developer workflows without creating additional complications.

@inetol what would you suggest to implement the developer workflows?

(For example, only performing the build after all tests have been completed.)

inetol replied:
We could do this under `themes.all` and check the lint there before building. I'm not too convinced on mixing different actions (like `themes.simple` having the linting process), as lint is *not always* a hard requirement to successfully build.

*themes* and *vite* wrappers are redundant and confusing rn, but are not the scope of this PR.

return42 commented on client/simple/src/js/main/preferences.ts:
```diff
@@ -0,0 +1,66 @@
+import { http, listen, settings } from "../core/toolkit.ts";
+
+const loadEngineDescriptions = async (): Promise<void> => {
+  let engineDescriptions: Record<string, [string, string]> | null = null;
```
In the old implementation the `http://127.0.0.1:8888/engine_descriptions.json` was requested only once .. with this new implementation this URL is requested on every `mouseenter` event (from below).

inetol replied:
d497509560f1551f38d3beac8c1c5a86de862af3

return42 commented on client/simple/src/js/main/preferences.js:
```diff
@@ -1,52 +0,0 @@
-  }
-
-  searxng.ready(() => {
-    let engine_descriptions = null;
-
-    function load_engine_descriptions() {
-      if (engine_descriptions == null) {
```
Related to https://github.com/searxng/searxng/pull/5073/files#r2256037965

Here (in this closure) the json from `http://127.0.0.1:8888/engine_descriptions.json` was stored in the variable `engine_descriptions` to avoid a reload on every `mouseenter` event.

return42 commented on client/simple/src/js/main/results.ts:
```diff
@@ -0,0 +1,178 @@
+  // set handlers in the on-properties
+  imgLoader.onload = (): void => {
+    onSuccess();
+  };
```
In the old implementation there was a spinner:

https://github.com/searxng/searxng/blob/6cccb46f2b2839daee13c5d12e86a7987a22e6cd/client/simple/src/js/main/results.js#L63-L65

was it intended to remove the spinner? .. for me it is OK / I just noticed that.

----

The spinner was shown as long the image was loaded:

https://github.com/searxng/searxng/blob/6cccb46f2b2839daee13c5d12e86a7987a22e6cd/client/simple/src/js/main/results.js#L119-L121

.. and removed when the load of the image finished or an error occurred:

https://github.com/searxng/searxng/blob/6cccb46f2b2839daee13c5d12e86a7987a22e6cd/client/simple/src/js/main/results.js#L80-L88

inetol replied:
Yes, I removed this because it looked ugly the way it was setup and there was layout shift (if I remember, there was also a 1sec cooldown before loading the "original" image, I don't know the reason for this, but all of that logic was removed)

return42 replied:
> Yes, I removed this because it looked ugly the way 

I agree, the spinner was ugly, but ..

> there was also a 1sec cooldown before loading the "original" image,

we need this delay back ..  When a user swipe from one image to the next, without staying on an image for even a second, then there is no need to load a high resolution for that image:

- https://github.com/searxng/searxng/pull/4198#issuecomment-2596278400
- https://github.com/searxng/searxng/issues/4214
- https://github.com/searxng/searxng/pull/4217

As far as I remember, the loading process was also aborted if the loading process had started (after the delay) but the user swiped to the next image while it was loading.

Long story short: Loading high resolution images is very resource-intensive, which is why we should always try not to request unnecessary ones.

----

Were any other features intentionally omitted from this PR? / *I'm trying to figure it out by comparing the new and old code & behavior, but it's not easy without knowing what to look for.*

inetol replied:
Done a8eccf69bd2678ae138aa2ab2aa16162c21c0ebe

Also fixed the full image if failed to load then keep the thumbnail (frequent with image proxy) and not fallback to the error image.

return42 commented on client/simple/src/js/main/results.ts:
```diff
@@ -0,0 +1,178 @@
+
+    if (backToTopElement && resultsElement) {
+      const scrollTop = document.documentElement.scrollTop || document.body.scrollTop;
+      const isScrolling = scrollTop >= 100;
+      resultsElement.classList.toggle("scrolling", isScrolling);
```
Not related to this PR (same issue in current master branch):  When [`infinite_scroll:`](https://docs.searxng.org/admin/settings/settings_ui.html) is active, the *back to top* button is never shown  / especially in *infinitely long* result lists such a button would be convenient.

update: *back to top* button is shown in the *desktop view* but not on smaller devices / below `(max-width:79.75em)`.

return42 commented on client/simple/src/js/main/search.ts:
```diff
@@ -0,0 +1,232 @@
+
+// autocompleter
+if (settings.autocomplete) {
+  let timeoutId: number;
```
> Lay the foundation for loading scripts granularly depending on the endpoint it's on.

Should we move the code block in this  condition (and the `fetchResults` from line 30 above) into a `autocomplete.ts` script?

inetol replied:
Done a8eccf69bd2678ae138aa2ab2aa16162c21c0ebe

return42 committed:
[license] client/simple: SPDX-License-Identifier: AGPL-3.0-or-later

SPDX short-form identifiers to communicate license information in a simple,
efficient, portable and machine-readable manner [1]

[1] https://spdx.dev/learn/handling-license-info/

return42 committed:
[fix] drop obsolte client/simple/static/

return42 committed:
[fix] add missing nvm environment to the "./manage dev.env" command

return42 committed:
[build] /static

return42 reviewed (approved):
I pushed 3 commits with some nits on top and I modified the commit to have all modification on files in folder `searx/static` in a final `[build] /static` commit.

---

There is an issue in the CI 

- https://github.com/searxng/searxng/actions/runs/17038728413/job/48297480603?pr=5073#step:6:764

I guess its not related to this PR .. but I will have a look to see whats wrong with this unit test .. could take a while, I can't reproduce the issue on my desktop!?!

return42 commented on searx/templates/simple/base.html:
```diff
@@ -2,27 +2,28 @@
-  <!--[if gte IE 9]>-->
-  <script src="{{ url_for('static', filename='js/searxng.head.min.js') }}" client_settings="{{ client_settings }}"></script>
-  <!--<![endif]-->
+  <script>
+    // update the css
+    document.documentElement.classList.remove('no-js');
+    document.documentElement.classList.add('js');
+  </script>
```
The inline script needs to be removed from the HTML, see

- https://github.com/searxng/searxng-docker/pull/424
